docs/0002-events-architecture.md

```diff
@@ -0,0 +1,84 @@
+# 2. Events Architecture
+
+Date: 2024-06-13
+
+## Status
+
+Proposed
+
+## Context
+
+Currently, in the Visual Editor, Google Analytics (GA) tracking is implemented by hardcoding event data attributes directly into the component. This approach has limitations:
+
+- It tightly couples the GA tracking logic with the Visual Editor.
+- Any changes to GA tracking requirements necessitate changes in the Visual Editor component code.
+- The need to pass extra attributes for GA event data complicates the component logic.
+
+To address these issues, we propose implementing an Event Emitter system within the Visual Editor. This will allow us to decouple the GA tracking logic from the Visual Editor component. 
+Events will be emitted when users interact with the component (e.g., clicking a button or selecting an option from a dropdown). 
+These events can then be listened to by any part of the application to trigger GA data layer pushes or other actions, without requiring changes to the Visual Editor component itself.
+
+## Decision
+
+We will implement an Event Emitter system in the Visual Editor. The component will emit events for user interactions such as button clicks and dropdown selections. 
+These events will carry necessary data attributes, and other parts of the application can listen to these events to perform actions like GA tracking.
+
+The following steps outline the implementation:
+
+1. Create an `EventEmitter` class to manage event listeners and emit events.
+2. Integrate the `EventEmitter` into the `VisualEditor` component.
+3. Emit events from the `VisualEditor` component when user interactions occur (e.g., `clickButton`, `selectOption`).
+4. Other parts of the application can register event listeners to handle these events and push data to the GA data layer as needed.
+
+The proposed class diagram is as follows:
+
+```mermaid
+classDiagram
+    class EventEmitter {
+        +events: Object
+        +on(event: String, listener: Function): void
+        +off(event: String, listenerToRemove: Function): void
+        +emit(event: String, data: Object): void
+    }
+
+    class VisualEditor {
+        +clickButton(): void
+        +selectOption(): void
+    }
+    
+    EventEmitter <|-- VisualEditor
+    
+    class User {
+        +visualEditor: VisualEditor
+        +onClickButton(data: Object): void
+        +onSelectOption(data: Object): void
+    }
+    
+    VisualEditor --> "1..*" EventEmitter : emits
+    User --> "1" VisualEditor : listens to
+    
+    EventEmitter : +events
+    EventEmitter : +on(event, listener)
+    EventEmitter : +off(event, listenerToRemove)
+    EventEmitter : +emit(event, data)
+    
+    VisualEditor : +clickButton()
+    VisualEditor : +selectOption()
+    
+    User : +onClickButton(data)
+    User : +onSelectOption(data)
+```
```

Comment:
Can you provide a usage example for the above classes, e.g. for handling a button click event?

Reply:
mmm 🤔 , I could try, but that will be getting too much into implementation detail i think...

Reply:
I personally find the diagram pretty confusing: 
1. The arrows are not clear about the relationship dynamics
2. "User" should probably be "Publishing App", but that also makes it look like EventEmitter is a separate application? 

I wonder if it's worth putting a placeholder for implementation detail until we actually implement it and then put in real code snippets and refine the diagram for then? 

+1 to Ryan - would be nice for it to be surrounding a specific example.